### DISCUSSION TITLE
Time Series Models | Bias Variance Tradeoffs

### DISCUSSION THREAD

**Discussion by ngupta23** [pycaret/pycaret#1906] [FAQs]
# Background

When building a model for any machine learning problem, we must pay attention to bias-variance tradeoffs. For time series: 

1. Having too much bias (underfitting) in the model means that the model is not able to capture all the signals in the data and hence leads to a higher error during the training phase (and in turn during the prediction phase as well).
2. Having too much variance in the model (overfitting) in the model means that the model is not able to generalize well to unseen future data (i.e. it can do well on the training data but is not able to predict as well as it trained). 

Let's see how this can be diagnosed with PyCaret.

# Solution in pycaret

## Overview 

The solution in pycaret is based on the recommendation in **[this video](https://www.youtube.com/watch?v=MyBSkmUeIEs)** by Andrew Ng. I would highly recommend the readers go through this video first before continuing with this article.

The steps followed in pycaret are:

1. The time-series data is first split into a train and test split.
2. The train split is then cross-validated across multiple folds.
3. The cross-validation error is used to select from multiple models during the hyperparameter tuning phase.
4. The final generalization error can then be reported using the test set.

## Setup
We will use the classical "airline" dataset to demonstrate this. A notebook for this article can be **[found here](https://github.com/ngupta23/medium_articles/blob/main/time_series/pycaret/pycaret_bias_variance_tradeoff.ipynb)**

```python
from pycaret.internal.pycaret_experiment import TimeSeriesExperiment
from pycaret.datasets import get_data

#### Get data ----
y = get_data("airline", verbose=False)

#### Setup Experiment ----
exp = TimeSeriesExperiment()

#### Forecast Horizon = 12 months and 3 fold cross validation ----
exp.setup(data=y, fh=12, fold=3, session_id=42)
```

## Train-Test Split

To diagnose bias-variance tradeoffs, `pycaret` initially splits the time series data into a train and test split. The temporal dependence of the data is maintained when performing this split. The length of the test set is the same as the forecast horizon that is specified while setting up the experiment (12 in this example). This is shown in the setup summary.

 
![image](https://user-images.githubusercontent.com/33585645/144029533-24b113ee-e940-4e8a-a490-e5953ddd3cd7.png)

This split can also be visualized using the `plot_model` functionality.

```python
exp.plot_model(plot="train_test_split")
```
![image](https://user-images.githubusercontent.com/33585645/144029160-e5b8d224-a545-49a8-86c5-a9d92606b916.png)

## Cross-Validation

Next, the training split is broken into cross-validation folds. This is done so that the training is not biased by just one set of training data. For example, if the last 12 months in the data were an anomaly (say due to Covid), that might impact the performance of an otherwise good model. Alternately, it might make an otherwise bad model look good. We want to avoid such situations. Hence, we train multiple times over different datasets and average the performance.

It is important to again maintain the temporal dependence when training across these multiple datasets (also called "folds"). Many strategies can be applied like expanding or sliding window strategies. More information about this can be **[found here](https://github.com/pycaret/pycaret/discussions/1761)**.

The number of folds can be controlled during the setup stage. By default `pycaret` time series module uses 3 folds. The folds in the training data can also be visualized using `plot_model` again. The blue dots represent the time points that are used for training in each fold and the orange dots represent the time points used to test the performance of the training fold. Again, the length of the orange dots is the same as the forecast horizon (12)

```python
exp.plot_model(plot="cv")
```

![image](https://user-images.githubusercontent.com/33585645/144031547-ddea4632-f258-4a27-80c9-5bcbea68ee28.png)

## Creating Initial Model

We will take the example of a reduced regression model for this article. More information about reduced regression models can be **[found here](https://github.com/pycaret/pycaret/discussions/1760)**.

```python
model = exp.create_model("lr_cds_dt")
```

![image](https://user-images.githubusercontent.com/33585645/144032798-73e3839a-f22f-4a9d-a0c8-cf4aa01353ec.png)

The performance is displayed across the 3 folds. The average MAE across the 3 folds is > 30 and the mean absolute percentage error (MAPE) is > 8%. Depending on the application, this may not be good enough.

Once the cross-validation is done, `pycaret` returns the model that is training on the entire training split. This is done so that the generalization of the model can then be tested on the test dataset that we held out previously.

```python
exp.predict_model(model)
exp.plot_model(model)
```
![image](https://user-images.githubusercontent.com/33585645/144033196-50b88392-bdda-46ed-ad1f-b8156983287e.png)

Prediction using this model now shows the forecasts for the time points corresponding to the test dataset. The metrics for these "test" forecasts are also displayed. These are worse than the metrics obtained during the cross-validation. Since the metrics were bad to begin with (high cross-validation errors), this is indicative of a high bias in the model (i.e. the model is not able to capture the trends in the dataset well at this point) as well as high variance (see video from Andrew Ng above). This is also visible in the prediction plot which shows that the blue prediction line is not close to the corresponding black line during the test period.

Let's see if we can improve the bias by tuning the model's hyperparameters.

## Tuning the Model to Improve Performance

Tuning the model tries various hyperparameter combinations to see which ones fit the model the best. Once the various hyperparameter combinations are tried out, the best hyperparameters are picked based on the mean cross-validation error. The cross-validation metrics using these best hyperparameters are then displayed.

```python
tuned_model = exp.tune_model(model)
```
![image](https://user-images.githubusercontent.com/33585645/144033892-31e80a6a-7a20-430b-9328-af8d1a8a574e.png)

So, we have been able to reduce the errors during the cross-validation stage to approximately <= 20 and the MAPE is < 5% by performing hyper-parameter tuning. This is much better than before. But now we need to ensure we are not overfitting the data. Let's look at the performance across the test dataset again.

```python
exp.predict_model(tuned_model)
exp.plot_model(tuned_model)
```

![image](https://user-images.githubusercontent.com/33585645/144034256-06d7488f-a461-47b3-831c-5166586e5d7d.png)

The forecasts across the test dataset show better performance than the cross-validation metrics indicative of a lack of overfitting. The plot also shows a good match to the actual data points during the test period.

So this model looks good. But what we need is the ability to predict the true "unknown" future data. This can be done by finalizing the model. 

## Finalizing the Model to make Future Predictions

```python
final_model = exp.finalize_model(tuned_model)
```

**Finalizing the model takes the hyperparameters from the model (tuned_model in this case), and fits the entire dataset using these hyperparameters.**

```python
print(exp.predict_model(final_model))
exp.plot_model(final_model)
```

![image](https://user-images.githubusercontent.com/33585645/144035101-5f948a95-c5d6-4aa1-8aec-ecafbed8d0a1.png)

And there you have it. Our best model is now able to make future predictions. 

That's it for this article. Until next time, Happy Forecasting!

**Comment by moezali1** [comment]
PyCaret 4.0 sunset the legacy time-series module (sktime/prophet integration), so this FAQ no longer maps to a supported code path.

If you still need time-series forecasting, use sktime / prophet / statsforecast directly — those projects have surpassed what the PyCaret wrapper offered.

For 4.0 docs, see https://pycaret.org/docs and the cleanup notice at https://github.com/pycaret/pycaret/issues/4177.

Closing as outdated.
